src/uct/sm/mm/mm_ep.c

```diff
@@ -58,10 +84,76 @@ ucs_status_t uct_mm_ep_put_bcopy(uct_ep_h tl_ep, uct_pack_callback_t pack_cb,
     return UCS_OK;
 }
 
-ucs_status_t uct_mm_ep_am_short(uct_ep_h ep, uint8_t id, uint64_t header,
+inline uint64_t uct_mm_ep_head_index_in_fifo(uct_mm_ep_t *ep)
+{
+	uct_mm_iface_t *iface = ucs_derived_of(ep->super.super.iface, uct_mm_iface_t);
+    return (ep->fifo_ctl->head & iface->fifo_mask);
+}
+
+uct_mm_fifo_element_t* uct_mm_ep_get_remote_elem(uct_mm_ep_t *ep, uint64_t head)
+{
+    uct_mm_iface_t *iface = ucs_derived_of(ep->super.super.iface, uct_mm_iface_t);
+    uct_mm_fifo_element_t* write_elem;  /* the fifo_element which the head points to */
+    uint64_t elem_index;                /* the fifo elem's index in the fifo. */
+								        /* must be smaller than fifo size */
+    uint64_t returned_val;
+
+    elem_index = uct_mm_ep_head_index_in_fifo(ep);
+    write_elem = (uct_mm_fifo_element_t*)(ep->fifo + elem_index * (iface->elem_size));
+
+    /* try to get ownership on the head element */
+    returned_val = ucs_compare_and_swap(&ep->fifo_ctl->head, head, head+1);
+    if (returned_val != head) {
+        return NULL;
+    }
+
+    return write_elem;
+}
+
+ucs_status_t uct_mm_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t header,
                                 const void *payload, unsigned length)
 {
-    return UCS_ERR_UNSUPPORTED;
+	uct_mm_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_mm_iface_t);
+	uct_mm_ep_t *ep = ucs_derived_of(tl_ep, uct_mm_ep_t);
+	uct_mm_fifo_element_t *elem_to_write;
+	uint64_t head;
+
+	UCT_CHECK_AM_ID(id);
+	UCT_CHECK_LENGTH(length + sizeof(header),
+					 iface->elem_size -	sizeof(uct_mm_fifo_element_t), "am_short");
+
+	head = ep->fifo_ctl->head;
+	/* check if there is room in the remote process's receive fifo to write */
+	if ((head - ep->fifo_ctl->tail) >= iface->config.fifo_size) {
+		return UCS_ERR_NO_RESOURCE;
+	}
+
+	elem_to_write = uct_mm_ep_get_remote_elem(ep, head);
+
+	if (elem_to_write != NULL ) {
+		/* write to the remote fifo */
+		*(uint64_t*)(elem_to_write + 1) = header;
+		memcpy((void*)(elem_to_write + 1) + sizeof(header), payload, length);
```

Comment:
Not urgent, but some thoughts - bcopy will look almost identical with deference that callback is invoked instead of memcpy call. In long term probably you will want convert some the initialization to inline func or macro. I guess you can rework it once bcopy is implemented